tests/test_authority.py

```diff
@@ -123,3 +128,102 @@ class MockResponse(object):
+
+    def test_validate_authority_boolean_should_skip_validation_and_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        """Pending deprecation, but kept for backward compatibility, for now"""
+        app = msal.ClientApplication(
+            "id", authority="https://contoso.com/common", validate_authority=False)
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_not_called()
+
+    def test_by_default_adfs_should_skip_validation_and_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        """Not strictly required, but when/if we already supported it, we better keep it"""
+        app = msal.ClientApplication("id", authority="https://contoso.com/adfs")
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_not_called()
+
+    def test_by_default_b2c_should_skip_validation_and_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        """Not strictly required, but when/if we already supported it, we better keep it"""
+        app = msal.ClientApplication(
+            "id", authority="https://login.b2clogin.com/contoso/b2c_policy")
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_not_called()
+
+    def test_known_authorities_should_allow_multiple_idempotent_calls_but_no_changes(self, *mocks):
+        known_hosts = ["contoso.com", "fabricam.com"]
+        msal.ClientApplication.set_known_authority_hosts(known_hosts)
+        msal.ClientApplication.set_known_authority_hosts(known_hosts)  # Allows idempotent calls
+        msal.ClientApplication.set_known_authority_hosts(
+            # Optional. Here we treat hosts orderless therefore accept equivalent call
+            reversed(known_hosts))
+        with self.assertRaises(ValueError):
+            msal.ClientApplication.set_known_authority_hosts(["hacked.com"])
+
+    def test_known_authority_hosts_should_not_coexist_with_validate_authority_boolean(self, *mocks):
+        msal.ClientApplication.set_known_authority_hosts(["private.cloud"])
+        with self.assertRaises(ValueError):
+            msal.ClientApplication(
+                "id",
+                authority="https://unknown.com/common", validate_authority=False)
+
+    def test_known_to_developer_authority_should_skip_validation_and_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        msal.ClientApplication.set_known_authority_hosts(["private.cloud"])
+        app = msal.ClientApplication("foo", authority="https://private.cloud/foo")
```

Comment:
Is this the expected code pattern for most / all cloud apps and tools? As it would be quite regressive for a "developer" to be building a tool which only works in public clouds / doesn't work in air-gapped or private-cloud environments. Take Azure CLI for example. The devs of that tool would have the user pass-in the authority endpoint for the cloud they are connecting to. And they can't know whether any of these features are supported by the target cloud, so they'd just take the user-provided input and set it as a known authority? Is that right? Or is the idea they'd leave the known authorities collection empty and expose separate flags for enabling authority validation, instance discovery, and alias lookups? But how would the "user" know whether to enable them or disable them? Or, would Azure CLI devs just always disable them entirely?

Reply:
> Is this the expected code pattern for most / all cloud apps and tools?

In short, yes, but that new optional `set_known_authority_hosts()` call is only needed when a downstream app is deployed to work with a small set of known (-to-developer) authorities. To give you a concrete example, the first four test cases in this PR do not need that `set_known_authority_hosts()` call.

> As it would be quite regressive for a "developer" to be building a tool which only works in public clouds / doesn't work in air-gapped or private-cloud environments. Take Azure CLI for example. The devs of that tool would have the user pass-in the authority endpoint for the cloud they are connecting to. And they can't know whether any of these features are supported by the target cloud, so they'd just take the user-provided input and set it as a known authority? Is that right?
> Or is the idea they'd leave the known authorities collection empty and expose separate flags for enabling authority validation, instance discovery, and alias lookups? But how would the "user" know whether to enable them or disable them? Or, would Azure CLI devs just always disable them entirely?

First of all, there is no longer such a thing as "(whether) they can't know whether *any of these features (enabling authority validation, instance discovery, and alias lookups)* are supported by the target cloud". Based on the simplified proposal, here this implementation PR exposes only one conceptual setting "what are the authorities which are unknown to Microsoft, but *you* still allow/want this app to operate with". It is a pure conceptual declaration, "I'm deploying this app to my contoso.com so I tell this app to proceed with authority contoso.com", and then MSAL will figure out the rest (which is basically skip all under-the-hood checks).

And yes, a MSAL-powered tool needs to expose such a new setting, when and only when the tool developer envisions their tool to be potentially deployed into a private cloud. The actual logistics is decided by each tool. Take Azure CLI for example, their ["az cloud" commands](https://docs.microsoft.com/en-us/cli/azure/cloud?view=azure-cli-latest) could be considered as a strong signal that the end user is aware of, and explicitly opts in to use an unusual cloud. So, that cloud could be fed into MSAL's new `set_known_authority_cloud()`. On the other hand, an app somehow encountered a "phishing.fabricam.com" authority during runtime, should not blindly set it as a "known authority" and subsequently bring end user to that compromised website and input their username and password there.

> Or, would any devs just always disable the 3 sub-features entirely? By always setting any "unreliable.fabricam.com" as a known authority?

I rephrased your question to make it more clear. The answer is:

* They don't need to. See the paragraph above.
* But if they insist, we cannot prevent them from doing so. Just keep in mind that, this defeats the purpose of authority validation, which is a security feature.
* But at least our new declarative `set_known_authority_hosts(...)` makes a developer less likely to "always disable authority validation", because they won't be able to enumerate all "phishing.fabricam.com", "hacked.contoso.com" etc. beforehand. (FWIW, in older version of MSALs, there used to be a boolean flag to simply turn off authority validation regardless of the incoming host name. That boolean flag could potentially be misused, unknowingly (pun not intended).)

Reply:
UPDATE: We worked out #496 to give a decisive boolean flag to turn off Instance Discovery endpoint I/O.